File: examples/rust/table/saga-hand/src/splitter_example.rs
//! Saga splitter pattern example for documentation.
//!
//! Demonstrates the splitter pattern where one event triggers commands
//! to multiple different aggregates.

use angzarr_client::proto::angzarr::{CommandBook, CommandPage, Cover, Uuid};
use angzarr_client::proto::examples::{TableSettled, TransferFunds};
use angzarr_client::SagaContext;
use prost_types::Any;

// docs:start:saga_splitter
fn handle_table_settled(event: &TableSettled, context: &SagaContext) -> Vec<CommandBook> {
    // Split one event into commands for multiple player aggregates
    event.payouts.iter().map(|payout| {
        let cmd = TransferFunds {
            table_root: event.table_root.clone(),
            amount: payout.amount.clone(),
        };

        let target_seq = context.get_sequence("player", &payout.player_root);

        CommandBook {
            cover: Some(Cover {
                domain: "player".into(),
                root: Some(Uuid { value: payout.player_root.clone() }),
                ..Default::default()
            }),
            pages: vec![CommandPage {
                sequence: Some(angzarr_client::proto::angzarr::command_page::Sequence::Num(target_seq)),
                command: Some(Any::from_msg(&cmd).unwrap()),
                ..Default::default()
            }],
            ..Default::default()
        }
    }).collect() // One CommandBook per player
}
// docs:end:saga_splitter

File: examples/python/table/saga-hand/splitter_example.py
"""Saga splitter pattern example for documentation.

Demonstrates the splitter pattern where one event triggers commands
to multiple different aggregates.
"""

from angzarr_client import SagaContext
from angzarr_client.proto.angzarr import types_pb2 as types
from angzarr_client.proto.examples import table_pb2 as table
from angzarr_client.proto.examples import player_pb2 as player


# docs:start:saga_splitter
def handle_table_settled(
    event: table.TableSettled, context: SagaContext
) -> list[types.CommandBook]:
    """Split one event into commands for multiple player aggregates."""
    commands = []

    for payout in event.payouts:
        cmd = player.TransferFunds(
            table_root=event.table_root,
            amount=payout.amount,
        )

        target_seq = context.get_sequence("player", payout.player_root)

        commands.append(
            types.CommandBook(
                cover=types.Cover(
                    domain="player", root=types.UUID(value=payout.player_root)
                ),
                pages=[types.CommandPage(sequence=target_seq, command=pack_any(cmd))],
            )
        )

    return commands  # One CommandBook per player


# docs:end:saga_splitter


def pack_any(msg):
    """Helper to pack proto message into Any."""
    from google.protobuf.any_pb2 import Any

    any_msg = Any()
    any_msg.Pack(msg)
    return any_msg
